websites/views.py

```diff
@@ -89,10 +107,101 @@ def get_queryset(self):
+
+
+class WebsiteCollaboratorViewSet(
+    NestedViewSetMixin,
+    viewsets.ModelViewSet,
+):
+    """ Viewset for Website collaborators along with their group/role """
+
+    serializer_class = WebsiteCollaboratorSerializer
+    permission_classes = (HasWebsiteCollaborationPermission,)
+    pagination_class = DefaultPagination
+    lookup_field = "username"
+
+    def get_queryset(self):
+        """
+        Get a list of all the users with permissions for this website, and annotate by group-name/role
+        (owner, administrator, editor, or global administrator)
+        """
+        website = get_object_or_404(
+            Website, name=self.kwargs.get("parent_lookup_website")
+        )
+        website_groups = list(
+            get_groups_with_perms(website).values_list("name", flat=True)
+        ) + [constants.GLOBAL_ADMIN]
+        owner_username = website.owner.username if website.owner else None
+
+        # Return the individual user and group if a primary key is provided
+        user_name = self.kwargs.get("username", None)
+        if user_name:
+            if user_name == owner_username:
+                return User.objects.filter(username=user_name).annotate(
+                    group=Value(constants.ROLE_OWNER, CharField())
+                )
+            group_subquery = Group.objects.filter(
+                Q(user__username=OuterRef("username")) & Q(name__in=website_groups)
+            )
+            return User.objects.filter(
+                Q(username=user_name) & Q(groups__name__in=website_groups)
+            ).annotate(group=Subquery(group_subquery.values("name")[:1]))
+
+        # Otherwise get all the collaborators and annotate with the relevant group they are in
+        query = User.objects.filter(username=owner_username).annotate(
+            group=Value(constants.ROLE_OWNER, CharField())
+        )
+        for group_name in website_groups:
+            query = query.union(
+                Group.objects.get(name=group_name)
+                .user_set.exclude(username=owner_username)
+                .annotate(group=Value(group_name, CharField()))
+            )
+        return query.order_by("name")
+
+    def destroy(self, request, *args, **kwargs):
+        """Remove the user from all groups for this website"""
+        instance = self.get_object()
+        website = Website.objects.get(name=self.kwargs.get("parent_lookup_website"))
+        for group in get_groups_with_perms(website):
+            instance.groups.remove(group)
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    def create(self, request, *args, **kwargs):
+        """ Add a user to the website as a collaborator"""
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        website = Website.objects.get(name=self.kwargs.get("parent_lookup_website"))
+        group_name = f"{ROLE_GROUP_MAPPING[serializer.validated_data.get('role')]}{website.uuid.hex}"
```

Comment:
Can we refactor this into a utility function? I see this pattern in `update` below and also a couple places in `websites/permissions.py`